src/core/api/track_choice_manager.ts

```diff
@@ -968,6 +1015,49 @@ export default class TrackChoiceManager {
 
     recursiveUpdateVideoTrack(0);
   }
+
+  /**
+   * Set the adaptation for stream.
+   * If the trick mode is enabled, select the trick mode track if it
+   * exists. If it does not exists, then disable the trick mode before
+   *
+   * switching adaptation.
+   * Select the main given adaptation if trick mode is disabled.
+   * @param {Object} adaptation$
+   * @param {Object | null} adaptation
+   */
+  private _setVideoTrack(adaptation$: Subject<Adaptation | null>,
+                         adaptation: Adaptation | null,
+                         switchTrickModeTo?: boolean): void {
+    if (adaptation === null) {
+      adaptation$.next(null);
+      return;
+    }
+    const trickModeEnabled = switchTrickModeTo ?? this.trickModeEnabled;
+    if (trickModeEnabled) {
+      if (adaptation.trickModeTracks !== undefined &&
+          adaptation.trickModeTracks[0] !== undefined) {
+        if (!this.trickModeEnabled) {
+          this.trickModeEnabled = true;
+          if (this.onTrickModeEnableChange) {
+            this.onTrickModeEnableChange(this.trickModeEnabled);
+          }
+        }
+        adaptation$.next(adaptation.trickModeTracks[0]);
+        return;
+      } else {
+        log.warn("TrackChoiceManager: No trickMode for video track. " +
+                 "Playing video track instead.");
+      }
+    }
+    if (this.trickModeEnabled) {
+      this.trickModeEnabled = false;
+      if (this.onTrickModeEnableChange) {
+        this.onTrickModeEnableChange(false);
```

Comment:
So we call the callback (which can call into the application) before actually changing the track.

I find it a little too risky.
I always prefer when we call into the API last.